rpc/RPCTransaction.java

```diff
@@ -165,12 +157,19 @@ public void close() {
             final UUID requestId = UUID.randomUUID();
             request.setId(requestId.toString());
             request.setLatencyMillis(networkLatencyMillis);
-            collectors.put(requestId, responseCollector);
+            addCollectorIfTransactionIsOpen(requestId, responseCollector);
             final QueryIterator<T> queryIterator = new QueryIterator<>(request.build(), requestObserver, responseCollector, transformResponse);
             return StreamSupport.stream(((Iterable<T>) () -> queryIterator).spliterator(), false);
         }
     }
 
+    private void addCollectorIfTransactionIsOpen(UUID requestId, ResponseCollector collector) {
```

Comment:
We use `compute` to atomically read our `AtomicBoolean` and add the new `ResponseCollector`.